framework/src/base/NonlinearSystem.C

```diff
@@ -1056,6 +1060,13 @@ NonlinearSystem::constraintResiduals(NumericVector<Number> & residual, bool disp
 
       if (constraints_applied)
       {
+        // If any of the above constraints inserted values in the residual, it needs to be assembled
+        // before adding the cached residuals below.
+        if ( residual_has_inserted_values )
+        {
+          residual.close();
```

Comment:
All we need is an assemble here without ghosting (because ghosting can happen after the addCachedResidualDirectly). But I couldn't find a way to assemble without ghosting.